libvast/src/transform_steps/projection.cpp

```diff
@@ -0,0 +1,129 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/transform_steps/projection.hpp"
+
+#include "vast/error.hpp"
+#include "vast/logger.hpp"
+#include "vast/plugin.hpp"
+#include "vast/table_slice_builder_factory.hpp"
+#include "vast/type.hpp"
+
+#include <arrow/type.h>
+#include <caf/expected.hpp>
+
+#include <algorithm>
+#include <utility>
+
+namespace vast {
+projection_step::projection_step(std::vector<std::string> fields)
+  : fields_(std::move(fields)) {
+}
+
+caf::expected<std::pair<vast::type, std::vector<int>>>
+projection_step::adjust_layout(const vast::type& layout) const {
+  auto to_keep = std::unordered_set<vast::offset>{};
+  auto flat_index_to_keep = std::vector<int>{};
+  const auto& layout_rt = caf::get<record_type>(layout);
+  for (const auto& key : fields_) {
+    auto offsets = layout_rt.resolve_key_suffix(key);
+    for (const auto& offset : offsets) {
+      if (to_keep.emplace(offset).second) {
+        flat_index_to_keep.emplace_back(layout_rt.flat_index(offset));
+      }
+    }
+  }
+  if (to_keep.empty())
+    return caf::no_error;
+  // adjust layout
+  auto transformations = std::vector<record_type::transformation>{};
+  for (const auto& [field, offset] : layout_rt.leaves()) {
+    if (!to_keep.contains(offset))
+      transformations.push_back({offset, record_type::drop()});
+  }
+  auto adjusted_layout_rt = layout_rt.transform(std::move(transformations));
+  if (!adjusted_layout_rt)
+    return caf::make_error(ec::unspecified, "failed to remove a field from "
+                                            "layout");
+  auto adjusted_layout = type{*adjusted_layout_rt};
+  adjusted_layout.assign_metadata(layout);
+  std::sort(flat_index_to_keep.begin(), flat_index_to_keep.end());
+  return std::pair{std::move(adjusted_layout), std::move(flat_index_to_keep)};
+}
+
+caf::expected<table_slice>
+projection_step::operator()(table_slice&& slice) const {
+  // adjust layout
```

Comment:
I don't think we need a comment that says "adjust layout" before calling a function called "adjust_layout". Either spell out the comment or remove it. In general comments should be full sentences that explain _why_ you're doing something.